Bal 1405 check alert def service

liorzam opened the pull request.

liorzblrn committed:
feat: run alerts

liorzblrn committed:
Merge branch 'alonp/feat/tm' into bal-1405-check-alert-def-service-v2

liorzblrn committed:
fix: seed

liorzblrn committed:
fix: seed

liorzblrn committed:
feat: run alerts

liorzblrn committed:
feat: run alerts

liorzblrn committed:
fix: align prisma schema with snake case 3

liorzblrn committed:
fix: tests using service

liorzblrn committed:
fix: data analytics service

liorzblrn committed:
fix: create alerts

liorzblrn committed:
chore: eslint

liorzblrn committed:
fix: pr fixes

liorzblrn committed:
fix(alert.service.ts): fix type casting of alertDefinition.inlineRule to InlineRule

The type casting of `alertDefinition.inlineRule` to `InlineRule` was incorrect. It was previously casted directly, which caused a compilation error. The fix involves introducing a new variable `unknownData` and casting it to `InlineRule` before assigning it to `inlineRule`. This ensures that the type casting is done correctly and eliminates the compilation error.

liorzam committed:
Merge branch 'alonp/feat/tm' into bal-1405-check-alert-def-service-v5

liorzblrn committed:
chore(migration.sql): remove unused migration file for updating alert state enum values
chore(migration.sql): update migration file for transaction monitoring
chore(schema.prisma): update enum values for AlertSeverity, AlertState, AlertStatus, AlertType, CounterpartyType, RiskCategory, ComplianceStatus, PEPStatus, and SanctionListMatchStatus
chore(generate-alerts.ts): refactor generateFakeAlertDefinition function to use async/await and improve readability

fix(seed.ts): rename generateFakeAlertDefinition to generateFakeAlertDefinitions for clarity and consistency
fix(alert.repository.ts): add exists method to check if an alert exists with given arguments and projectIds
fix(alert.service.ts): change AlertState and AlertStatus enum values to lowercase for consistency
feat(alert.service.ts): add support for skipping duplicate alerts during alert execution
feat(alert.service.ts): add logging and error handling during alert execution
feat(alert.consts.ts): add AlertExecutionStatus enum for alert execution status tracking

feat(prisma.service.ts): extend PrismaClient with custom methods exists and softDelete to provide additional functionality

The PrismaService class in prisma.service.ts has been extended to include custom methods exists and softDelete. The exists method checks if a record exists in the database based on the provided where condition. The softDelete method performs a soft delete by updating the deletedAt field of the record to the current date.

This extension provides additional functionality to the PrismaClient, allowing for more flexibility and control when working with the database.

Omri-Levy commented on services/workflows-service/src/alert/alert.service.ts:
```diff
@@ -103,33 +112,136 @@ export class AlertService {
-    // ...
-    return true;
+  private async checkAlert(alertDefinition: AlertDefinition): Promise<boolean> {
+    let inlineRule: InlineRule;
+
+    const unknownData: unknown = alertDefinition.inlineRule;
+
+    inlineRule = unknownData as InlineRule;
```
Why are these re-assignments and type changes needed?

liorzam replied:
i just got an ts issue about it
```
Conversion of type 'string | number | boolean | JsonObject | JsonArray | null' to type 'InlineRule' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Type 'JsonValue[]' is missing the following properties from type 'InlineRule': id, fnName, options, subjectsts(2352)
  ```

Omri-Levy commented on services/workflows-service/src/alert/alert.service.ts:
```diff
@@ -103,33 +112,136 @@ export class AlertService {
+        }
+      });
+
+    await Promise.all(alertsPromises);
```
Does this need to be `allSettled`?

liorzam replied:
nope the promise in handled since we have high level try..catch in the map function

Omri-Levy commented on services/workflows-service/src/alert/consts.ts:
```diff
@@ -0,0 +1,5 @@
+  SUCCEEDED: 'SUCCEEDED',
+  SKIPPED: 'SKIPPED',
+  FAILED: 'FAILED',
+};
```
```suggestion
} as const;
```

Omri-Levy commented on services/workflows-service/src/data-analytics/types.ts:
```diff
@@ -19,3 +25,11 @@ export type TransactionsAgainstDynamicRulesType = {
   groupByCounterparty?: boolean;
 };
+
+export type AnyArray = any[];
+
+export type AnyRecord = Record<PropertyKey, any>;
+
+export type GenericFunction = (...args: AnyArray) => any;
+
+export type GenericAsyncFunction = (...args: AnyArray) => Promise<any>;
```
Let's move these types to `workflows-service/src/types.ts` for now. I'll add a tech debt task to have some way to share these sort of types in our system.

liorzblrn committed:
chore(fetchers.ts): update alert severity, status, type, and state values to lowercase for consistency and improved readability

feat(fetchers.ts): update alert severity, status, type, and state values to lowercase for consistency and improved readability

fix(fetchers.ts): update alert severity, status, type, and state values to lowercase for consistency and improved readability

chore(generate-alerts.ts): update property name 'groupedBy' to 'subjects' for clarity and consistency

feat(generate-alerts.ts): update property name 'groupedBy' to 'subjects' for clarity and consistency

fix(generate-alerts.ts): update property name 'groupedBy' to 'subjects' for clarity and consistency

chore(alert-definition.repository.ts): update method name 'findFirst' to 'findFirstOrThrow' for clarity

feat(alert-definition.repository.ts): add 'findMany' method to retrieve multiple alert definitions

fix(alert-definition.repository.ts): update method name 'findFirst' to 'findFirstOrThrow' for clarity

chore(alert.service.ts): update log message to include alert definition ID in single quotes

feat(alert.service.ts): update log message to include alert definition ID in single quotes

fix(alert.service.ts): update log message to include alert definition ID in single quotes

feat(alert.service.ts): update error handling to include rejected alerts in the response

fix(alert.service.ts): update error handling to include rejected alerts in the response

chore(consts.ts): update AlertExecutionStatus values to lowercase for consistency

refactor(data-analytics.service.ts): remove unused imports and variables to improve code cleanliness
feat(data-analytics.service.ts): add logger dependency to the DataAnalyticsService constructor for better logging
feat(data-analytics.service.ts): replace console.log statements with logger.log and logger.debug for improved logging
refactor(types.ts): rename 'groupedBy' property in InlineRule type to 'subjects' for better semantics
refactor(types.ts): remove unused types AnyArray, AnyRecord, GenericFunction, and GenericAsyncFunction from types.ts

liorzblrn committed:
fix: pass ci

liorzblrn committed:
refactor(alert.controller.external.ts): remove unused imports and variables to improve code cleanliness and readability
refactor(alert.service.ts): change Boolean to boolean in the return type of isDuplicateAlert method for consistency and clarity

liorzblrn committed:
chore(data-analytics.service.intg.test.ts): import commonTestingModules from nest-app-helper to improve code organization and reusability

liorzblrn committed:
fix: status label
